### PR TITLE
Add support for special characters in snowflake identifiers for Snowflake Catalog

### PR DESCRIPTION
**Pull request by manisin**
Currently the catalog is unable to handle databases or schema or table names (snowflake identifiers) with special characters. This limitation is due to sanitizing of the parameter for the like clause of SQL statements (which is not well supported by Snowflake JDBC's PreparedStatement) to determine if a database or schema exists (eg: SHOW DATABASES LIKE '%s' IN ACCOUNT). The alternative proposal is to list sub-namespace (eg: list schemas within a database (SHOW SCHEMAS/TABLES IN IDENTIFIER(?) LIMIT 1) or list tables within schema) and rely on the exception being thrown if the database/schema would not exist. This avoids using the like statement (while still being performant) and allows passing in the snowflake identifiers with special characters as quoted identifiers (more details at https://docs.snowflake.com/en/sql-reference/identifiers-syntax.html). The table shows the corresponding commands for creating and accessing database/schema objects in snowflake and spark. The column "Quoted" refers to whether the identifier (database/schema name) is quoted as per snowflake's identifier convention.

```
+----+------------------------------------+-------------------+--------------------------------------------------+
| Row| Snowflake                          | Quoted?| Spark(3.3.1)                                                |
+----+------------------------------------+-------------------+--------------------------------------------------+
| 1  | create database "$peci@al"         | Yes    | spark.sql("use database `\"$peci@l\"`").show()                |
| 2  | create database """doubleQouted""" | Yes    | spark.sql("use database `\"\"\"doubleQuoted\"\"\""`).show()   |
| 3  | create database "Dot.ted"          | Yes    | spark.sql("use database `\"Dot.ted\"`").show()                |
| 4  | create schema "under_Score"        | Yes    | spark.sql("use database `\"Dot.ted\"`.`\"under_Score\"`").show()|
| 5  | create database lower              | No     | spark.sql("use database lower").show()                      |
| -  |                                    |        | spark.sql("use database LOWER").show()                      |
| -  |                                    |        | spark.sql("use database LoWeR").show()                      |
| 6  | create database dollar$            | No     | spark.sql("use database DOLLAR$").show()                    |
| -  |                                    |        | spark.sql("use database dollar$").show()                    |
| -  |                                    |        | spark.sql("use database DOllar$").show()                    |
| 7  | create database "lowerq"           | Yes    | spark.sql("use database `\"lowerq\"`").show()               |
+----+------------------------------------+--------+-------------------------------------------------------------+
```

### CODE REVIEW

**Review by nastra — COMMENTED** [review]
the changes mostly LGTM, but it's unclear why we're removing tests with special characters.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java [hunk 1]
```diff
@@ -203,31 +176,26 @@ public boolean schemaExists(SnowflakeIdentifier schema) {
...
     } catch (InterruptedException e) {
       throw new UncheckedInterruptedException(
           e, "Interrupted while checking if schema '%s' exists", schema);
     }
 
-    // Filter to handle the edge case of '_' appearing as a wildcard that can't be remapped the way
-    // it can for predicates in SELECT statements.
-    schemas.removeIf(sc -> !schema.schemaName().equalsIgnoreCase(sc.schemaName()));
-    return !schemas.isEmpty();
+    return true;
```
[the remark this answers is not in the document]
Comment (manisin) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java:198]: It could be possible that the schema exists without any tables being created yet.

File: snowflake/src/test/java/org/apache/iceberg/snowflake/JdbcSnowflakeClientTest.java [hunk 2]
```diff
@@ -79,46 +79,24 @@ public void testNullClientPoolInConstructor() {
...
 
     verify(mockQueryHarness)
         .query(
             eq(mockConnection),
-            eq("SHOW DATABASES LIKE 'DB_1' IN ACCOUNT"),
-            any(JdbcSnowflakeClient.ResultSetParser.class));
-  }
-
-  @Test
-  public void testDatabaseExistsSpecialCharacters() throws SQLException {
```
[the remark this answers is not in the document]
Comment (manisin) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/JdbcSnowflakeClientTest.java:95]: I have removed all the code that added custom massaging for the special characters by removing the use of LIKE clause.  The onus is on the user to pass the identifiers with special characters as quoted ( and escape/enclose them correctly) based on the syntax specified by snowflake's identifier documentation.

**Comment by manisin** [comment]
> the changes mostly LGTM, but it's unclear why we're removing tests with special characters.
Thanks for the review @nastra ! The test with special characters are removed as I also removed the code for custom handling of the special characters. The special characters have be escaped/quoted correctly according to spark/snowflake documentation by the user and they would be passed through the catalog library to snowflake without any modification.

**Comment by nastra** [comment]
would be good to also get feedback from @danielcweeks

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 3]
```diff
@@ -45,6 +45,7 @@ public class SnowflakeCatalog extends BaseMetastoreCatalog
     implements Closeable, SupportsNamespaces, Configurable<Object> {
   private static final String DEFAULT_CATALOG_NAME = "snowflake_catalog";
   private static final String DEFAULT_FILE_IO_IMPL = "org.apache.iceberg.io.ResolvingFileIO";
+  private static final String APP_IDENTIFIER = "iceberg-SDK";
```
Comment (danielcweeks) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:48]: This looks like it's unrelated to the special characters.  Can we pull these change out?  I'm ok with the rest of the PR, but I'm not sure we want to use this particular identifier and it's easier to have a discussion about that separately.

Comment (sfc-gh-mparmar) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java]: Removed the identifier.

Comment (manisin) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java]: Removed the application identifier, will create a separate PR for it.

**Review by danielcweeks — APPROVED** [review]
